src/lib.rs

```diff
@@ -910,44 +910,354 @@ pub mod contract {
+            let current_sqrt_price = pool.sqrt_price;
+            assert_eq!(pool.current_tick_index, init_tick);
+            assert_eq!(pool.sqrt_price, calculate_sqrt_price(init_tick).unwrap());
+
+            let position_amount = mint_amount - 1;
+
+            let liquidity_delta = get_liquidity_by_y(
+                TokenAmount(position_amount),
+                0,
+                MAX_TICK,
+                pool.sqrt_price,
+                false,
+            )
+            .unwrap()
+            .l;
+
+            let pool_key = PoolKey::new(token_x, token_y, fee_tier).unwrap();
+            let slippage_limit_lower = pool.sqrt_price;
+            let slippage_limit_upper = pool.sqrt_price;
+            create_position!(
+                client,
+                ContractRef,
+                dex,
+                pool_key,
+                0,
+                MAX_TICK,
+                liquidity_delta,
+                slippage_limit_lower,
+                slippage_limit_upper,
+                alice
+            );
+
+            Ok(())
+        }
+
+        #[ink_e2e::test]
+        async fn limits_big_deposit_and_swaps(mut client: ink_e2e::Client<C, E>) -> E2EResult<()> {
+            let (dex, token_x, token_y) =
+                init_dex_and_tokens_max_mint_amount!(client, ContractRef, TokenRef);
+
+            let mint_amount = 2u128.pow(76) - 1;
+            let alice = ink_e2e::alice();
+            approve!(client, TokenRef, token_x, dex, u128::MAX, alice);
+            approve!(client, TokenRef, token_y, dex, u128::MAX, alice);
+
+            let fee_tier = FeeTier {
+                fee: Percentage::from_scale(6, 3),
+                tick_spacing: 1,
+            };
+            create_fee_tier!(client, ContractRef, dex, fee_tier, alice);
+
+            let init_tick = 0;
+            create_pool!(
+                client,
+                ContractRef,
+                dex,
+                token_x,
+                token_y,
+                fee_tier,
+                init_tick
+            );
+
+            let pos_amount = mint_amount / 2;
+            let lower_tick = -(fee_tier.tick_spacing as i32);
+            let upper_tick = fee_tier.tick_spacing as i32;
+            let pool = get_pool!(client, ContractRef, dex, token_x, token_y, fee_tier).unwrap();
+            let liquidity_by_x = get_liquidity_by_x(
+                TokenAmount(pos_amount),
+                lower_tick,
+                upper_tick,
+                pool.sqrt_price,
+                false,
+            )
+            .unwrap()
+            .l;
+
+            let liquidity_by_y = get_liquidity_by_y(
+                TokenAmount(pos_amount),
+                lower_tick,
+                upper_tick,
+                pool.sqrt_price,
+                false,
+            )
+            .unwrap()
+            .l;
+
+            let liquidity_delta = if liquidity_by_y < liquidity_by_x {
```

Comment:
You also need to update this case:
1. Calculate liquidity for one of the tokens.
2. Calculate the amount of the other token based on the liquidity calculated in point 1.
3. Add similar validation to contract balances - they should be exactly the same as the assumed amounts in points 1 and 2.